### ISSUE TITLE
Slow initialization with nearest-projection mapping

### ISSUE THREAD

**Issue by threimann** [precice/precice#132]
The initialization with nearest-projection takes an unreasonable amount of time. I have experienced this several times. In my current example the setup is:

- serial explicit one-way coupling between FASTEST and Ateles
- coupling mesh is cubic and has either 32x32 or 128x128 nodes per side in FASTEST, and 32x32 nodes per side in Ateles
- FASTEST has 128 ranks, Ateles has 16 ranks, we compute on Sandy Bridge nodes

For the (6x)32x32 element mesh, preCICE spends 7min40s after the filtering (see output). In comparison, nearest-neighbor initializes almost instantly. On the 128x128 mesh, I never saw a successful initialization, even when giving Ateles 128 ranks as well (I canceled after 1hr waiting).

[ateles.out.txt](https://github.com/precice/precice/files/2179255/ateles.out.txt)
[precice.xml.txt](https://github.com/precice/precice/files/2179256/precice.xml.txt)

**Comment by uekerman** [comment]
At the core of this problem lies the fact that the nearest-projection mapping still has a sub-optimal quadratic complexity. Improvement to linear complexity would be possible, but is a decent work package, compare #99 

Let's look at the numbers:

- The FASTEST mesh increases by a factor of 16 from the small to the large mesh.
- The NP mapping has a complexity of O(N_F/P_A * N_A/P_A), where N_F and N_A are the sizes of both meshes and P_A is the number of ranks of Ateles at the interface.
- From the small to the large mesh, the compute time should, thus, increase by a factor of 16, which results in 122min40s (beyond your timeout). 
- Increasing the ranks of Ateles should have a positive influence. Let's assume that P_A doubles if the total number of ranks of Ateles is increased by 8 (from 16 to 128). The runtime should now decrease by a factor of 4 to roughly 30min. 

Apparently, the numbers do not coincide with your observation. What could go wrong?

- Maybe the assumption is wrong that the number of Ateles ranks at the interface doubles. I will ask Neda to comment on the domain decomposition of Ateles in such a case. 
- Maybe we run into a memory hierarchy problem. You could try to use:
```xml
<use-mesh name="AcousticSurfaceFASTEST" from="FASTEST" geometric-filter="filter-first"/>
```

The brute-force alternative is to use simply even more ranks for Ateles until we have resolved #99.

**Comment by threimann** [comment]
> The runtime should now decrease by a factor of 4 to roughly 30min.

I cannot say for sure that I waited this long for this attempt, so you could still be correct.

**Comment by g009266** [comment]
The increase of the ranks does not really mean, that all coupling elements will be computed by more/different processes. In our case the increase from 16 ranks to 128 ranks means, that the computational time should decrease by a factor of two, so that we end up at roughly 60 min. We also have imbalances in our simulations, which we also have to consider. I had a look at the partitioning of the domain when using 16/32/128 ranks for the Ateles domain. The cube represents the coupling interface to the FASTEST domain. When going over from 16 to 32 you can  not really see a difference, while with 128 there are around 2 times more ranks involved in the coupling (it is not very well visible in the picture, due to the very close colour scale). 
![proc16](https://user-images.githubusercontent.com/22452521/42516051-7d9e18dc-845d-11e8-87ac-b68d199e68b6.png)
![proc32](https://user-images.githubusercontent.com/22452521/42516052-7dd47c92-845d-11e8-87be-6c530a68eebb.png)
![proc128](https://user-images.githubusercontent.com/22452521/42516055-7e89e7f8-845d-11e8-8430-75fdbd4a212b.png)

**Comment by uekerman** [comment]
Thanks, @g009266. This is pretty much as I thought it is. The number of ranks of Ateles at the interface doubles when going from 16 to 128 total ranks. Please note that still we expect that the compute time decreases by a factor of 4 due to the quadratic complexity.

@threimann Could you please run again with 128 Ateles ranks? And maybe also with 256 and 512, just to see if this is a workaround for the time being. 

Filtering and the actual mapping are very similar operations here. It's, thus, not surprising to see similar timings (7 and 10mins). Still, of course, this makes the problem even worse.  

Bottom line: #99 is imperative.

**Comment by threimann** [comment]
Strangly enough, my last test on the 128 element FASTEST mesh with giving Ateles 128 ranks needed only 7s for the filtering and 6s for the initialization. While this is in a way good news, it is also not at all in agreement with your expectations. As far as I know, @g009266 has also reported inconsistent behavior with NP initialization. To me this frankly does not look like a pure scaling problem.

**Comment by threimann** [comment]
So sorry for the confusion. The above numbers (7s+6s) were obtained with NN. With NP we have 96min for the filtering, and 2min30s for the initialization.

[ateles128+128.out.txt](https://github.com/precice/precice/files/2187423/ateles128%2B128.out.txt)

**Comment by threimann** [comment]
A test with 128 Ateles ranks and `filter-first` option leaves us at 97min for pre-filtering, and 44s for initialization.

**Comment by fsimonis** [comment]
Closed by #156
